### ISSUE TITLE
Problem in rendering very large network scenario

### ISSUE THREAD

**Issue by igs3000** [usnistgov/NetSimulyzer#47]
Dear Evan Black

I am trying to visualize a Network topology of size [15000km x 15000km x 15000km] with 500 nodes (only simple spherical 3D models). I am simulating a 1-sec network scenario with no traffic at all.

It is said that NetSimulyzer can be able to simulate a network of any size.

But, even at any zoom level (through keyboard and mouse), I can not able to see the entire network scenario. Even I can not see more than one node at a time (by clicking the nodes from the list)

Even while zooming in one particular node, mostly a portion of the model is getting clipped.

I tried to scale down and scale up the size of individual node models (10m to 1000m), but a portion of the node model gets clipped or not at all rendered.

1) Is there any way to fit that huge network scenario (15000km x 15000km x 15000km); so that it can be visualized fully in the NetSimylyzer window?
2)  Further, generally, moving the camera using keyboard and mouse commands of the user interface is not at all practically optimal - I think, there should be some kind of automatic scaling of step values with respect to the size of the network topology (For example, the same is possible with Omnet++ visualization interface).
3) **Am I crossing the limits of the NetSimulyzer?**

**If possible give me some suggestions on altering some sections of code to make it work.**

Charles Pandian,

**Comment by bpe2** [comment]
Hey Charles,

It may be helpful if you provide a screenshot, but I think I know what's going on.

The camera has a (mostly) arbitrary limit for how far away it will show something, called the 'far plane' or 'clip plane'. Admittedly, the default right now is pretty low (~1000 ns-3 units). I intended to make this a configurable value, but right now, it's fixed at compile time. If you'd like to try raising this limit yourself without waiting for me to make this configurable, it's on [line 680 of `src/window/scene/SceneWidget.cpp`](https://github.com/usnistgov/NetSimulyzer/blob/master/src/window/scene/SceneWidget.cpp#LL680C112-L680C112), specifically the `1000.0f` is the value I'm referring to.

![Screenshot from 2023-05-23 09-52-11](https://github.com/usnistgov/NetSimulyzer/assets/54951512/bf54e4d9-0540-4bf6-a4d6-13a46287ddd4)

This limit is mostly for optimization, in most 3D applications it does not make sense to render everything all the time if the user can barely see it, but raising this limit to the point of practically removing it _shouldn't_ hurt. 

I'm particularly interested in your point 2 here. I'm not terribly familiar with Omnet++. Do you mean that it changes the distance scale it uses per display unit based on the overall size of the simulation, like if I had two scenarios each with two nodes, one where they're 500 m apart and another where they're 500 km apart, they would look the same, but there's some indication the scale has changed? That may be useful for a sparse but large scenario.

Let me know if changing that value accomplishes what you're after, or if I'm totally off base here.

Evan

**Comment by bpe2** [comment]
1. Comment out [Line 264 in  `src/window/scene/SceneWidget.cpp`](https://github.com/usnistgov/NetSimulyzer/blob/master/src/window/scene/SceneWidget.cpp#L286).
2. [Line 264 in `src/window/scene/SceneWidget.cpp`](https://github.com/usnistgov/NetSimulyzer/blob/master/src/window/scene/SceneWidget.cpp#L264) Change `glClearColor(1.0f, 1.0f, 1.0f, 1.0f);` to `glClearColor(0.0f, 0.0f, 0.0f, 1.0f);` and turn the skybox off in the settings
3. The `Height`, `Width`, and `Depth` attributes for [`NodeConfiguration` ](https://github.com/usnistgov/NetSimulyzer-ns3-module/blob/master/model/node-configuration.cc#L184) will accomplish this. You specify the height/width/depth in ns-3 units (meters, typically) and the application will calculate and apply that scale. This works with the `Scale` & `ScaleAxis` attributes, but I recommend you choose one or the other. This is bound by the [`KeepRatio` attribute](https://github.com/usnistgov/NetSimulyzer-ns3-module/blob/master/model/node-configuration.cc#L177) as well, so if you want a non-uniform scale, (not the same amount in all directions) then set that attribute to `false` as well

**Comment by bpe2** [comment]
For the movement speed of the camera, try increasing the 'Move Speed' setting in the 'Camera' section. The default is tuned around smaller scenarios. If the max value for that is too low, then that value can be manually set on [line 55 of `src/render/Camera.cpp`](https://github.com/usnistgov/NetSimulyzer/blob/master/src/render/camera/Camera.cpp#L55)

**Comment by igs3000** [comment]
Thank you, Evan.
**Everything worked well and as I expected.**

If possible, you may try to auto scale the camera movement step size in your future versions of NetSimulyzer.
It will be very helpful to manipulate the camera in the scenario irrespective of the topology size.

**Comment by bpe2** [comment]
Yeah, I agree with that. Maybe some sort of additional scale value based on the scenario size.

I admit, your scenario is much larger than what I'm used to. Thanks for bringing it to my attention. I guess I'll have to rethink what construes a 'large scenario'.

I'll break off each change into its own issue, and close this one.

**Comment by igs3000** [comment]
One more suggestion:
It will be easy to manipulate camera to move in and out using the center scroll button of the mouse.
If possible, you may add this functionality in in your future versions of NetSimulyzer.

Anyway thank you for the excellent support provided.

**Comment by bpe2** [comment]
I've thought about adding a 'trackball' style control, similar to how Blender works. Where there is a center point that the camera rotates around and zooms in and out using the mouse wheel. If you think something like that would be useful, I can add that as a todo item. It wouldn't be too challenging
